kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/ChecksumLogReplayMetricsTestBase.scala

```diff
@@ -0,0 +1,146 @@
+
+  /////////////////////////
+  // Test Helper Methods //
+  /////////////////////////
+
+  /**
+   * Creates a temporary directory with a test engine and builds a table with CRC files.
+   * Returns the created table and engine for testing.
+   *
+   * @param f code to run with the table and engine
+   */
+  protected def withTableWithCrc(f: (Table, String, MetricsEngine) => Any): Unit = {
+    withTempDirAndMetricsEngine { (path, engine) =>
+      // Produce a test table with 0 to 11 .json, 0 to 11.crc, 10.checkpoint.parquet
+      withSQLConf(DeltaSQLConf.DELTA_WRITE_CHECKSUM_ENABLED.key -> "true") {
+        spark.sql(
+          s"CREATE TABLE delta.`$path` USING DELTA AS " +
+            s"SELECT 0L as id")
+        for (_ <- 0 to 10) { appendCommit(path) }
+        assert(checkpointFileExistsForTable(path, 10))
+      }
+      val table = Table.forPath(engine, path)
+      f(table, path, engine)
+    }
+  }
+
+  // Abstract method to be implemented by concrete test classes
+  protected def loadSnapshotFieldsCheckMetrics(
+      table: Table,
+      engine: MetricsEngine,
+      expJsonVersionsRead: Seq[Long],
+      expParquetVersionsRead: Seq[Long],
+      expParquetReadSetSizes: Seq[Long],
+      expChecksumReadSet: Seq[Long],
+      readVersion: Long = -1): Unit
+
+  // Method to adjust list of versions of checkpoint file read.
+  protected def getExpectedCheckpointReadVersions(versions: Seq[Long]): Seq[Long] = versions
+
+  ///////////
+  // Tests //
+  ///////////
+
+  Seq(-1L, 0L, 3L, 4L).foreach { readVersion => // -1 means latest version
+    test(
+      s"checksum found at the read version: ${if (readVersion == -1) "latest" else readVersion}") {
+      withTableWithCrc { (table, _, engine) =>
+        loadSnapshotFieldsCheckMetrics(
+          table,
+          engine,
+          // shouldn't need to read commit or checkpoint files as P&M are found through checksum
+          expJsonVersionsRead = Nil,
+          expParquetVersionsRead = Nil,
+          expParquetReadSetSizes = Nil,
+          expChecksumReadSet = Seq(if (readVersion == -1) 11 else readVersion),
+          readVersion)
+      }
+    }
+  }
+
+  test(
+    "checksum not found at read version and checkpoint exists at read version => use checkpoint") {
+    withTableWithCrc { (table, tablePath, engine) =>
+      val checkpointVersion = 10
+      deleteChecksumFileForTable(tablePath, Seq(checkpointVersion))
+
+      loadSnapshotFieldsCheckMetrics(
+        table,
+        engine,
+        // 10.crc missing, 10.checkpoint.parquet exists.
+        // Attempt to read 10.crc fails and read 10.checkpoint.parquet succeeds.
+        expJsonVersionsRead = Nil,
+        expParquetVersionsRead = Seq(10),
+        expParquetReadSetSizes = getExpectedCheckpointReadVersions(Seq(1)),
+        expChecksumReadSet = Seq(10),
+        readVersion = 10)
+    }
+  }
+
+  test(
+    "checksum missing read version and the previous version, " +
```

Comment:
Can we improve the readability of these test cases a bit?

`checksum missing at read version & prev version; checkpoint exists at read version & prev version => use checkpoint`

Reply:
also -- what about the cross product of all the other cases?

- this case: checksum missing at X and X-1, checkpoint at X and X-1
- other case: checksum missing at X ONLY, checkpoint at X and X-1
- checksum EXISTS at X and X-1, checkpoint at X and X-1
- etc etc

Reply:
This one is checksum missing at read version & prev version; checkpoint exists at prev version => use checkpoint, updated naming

checksum missing at X ONLY, checkpoint at X is the one above "checksum not found at read version and checkpoint exists at read version => use checkpoint"

checksum EXISTS at X, we should use checksum anyway so it is a basic test case "checksum found at the read version xxxx"

checksum  EXISTS X-1 and checkpoint at X-1, I replicate the one in P&MLoading(using crc) suite to DMloading(using checkpoint) with a todo to merge to one after incremental loading being implemented